src/reference/asciidoc/kafka.adoc

```diff
@@ -2438,9 +2439,58 @@ This type inference can be achieved only when the `@KafkaListener` annotation is
 With a class-level `@KafkaListener`, the payload type is used to select which `@KafkaHandler` method to invoke, so it must already have been converted before the method can be chosen.
 ====
 
-NOTE: When you use the `StringJsonMessageConverter`, you should use a `StringDeserializer` in the Kafka consumer configuration and a `StringSerializer` in the Kafka producer configuration when you use Spring Integration or the `KafkaTemplate.send(Message<?> message)` method.
-When you use the `BytesJsonMessageConverter`, you should use a `BytesDeserializer` in the Kafka consumer configuration and `BytesSerializer` in the Kafka producer configuration when you use Spring Integration or the `KafkaTemplate.send(Message<?> message)` method (see <<kafka-template>>).
-Generally, the `BytesJsonMessageConverter` is more efficient because it avoids a `String` to and from `byte[]` conversion.
+[NOTE]
+====
+On the consumer side, you can configure a `JsonMessageConverter`; it can handle `ConsumerRecord` values of type `byte[]`, `Bytes` and `String` so should be used in conjunction with a `ByteArrayDeserializer`, `BytesDeserializer` or `StringDeserializer`.
+(`byte[]` and `Bytes` are more efficient because they avoid an unecessary `byte[]` to `String` conversion).
+You can also configure the specific subclass of `JsonMessageConverter` corresponding to the deserializer, if you so wish.
+
+On the producer side, when you use Spring Integration or the `KafkaTemplate.send(Message<?> message)` method (see <<kafka-template>>), you must configure a message converter that is compatible with the configured Kafka `Serializer`.
+
+* `StringJsonMessageConverter` with `StringSerializer`
+* `BytesJsonMessageConverter` with `BytesSerializer`
+* `ByteArrayJsonMessageConverter` with `ByteArraySerializer`
```

Comment:
Can't we have some built-in `CompositeSerializer`  to delegate to the appropriate Kafka `Serializer` according an incoming `payload` type? This way Io believe we just can stop to talk about serializers at all.

I understand that it might be a different story, but wanted to share with your a thought from my mind.
`

Reply:
This is for outbound. The super class handles all 3 inbound.

I don't know what you mean by "delegate" to serializer - that is called by Kafka.

Reply:
I don't talk about a hierarchy. I talk about a composition.
We just implement a simple `Serializer` and do a `Map.get()` logic based on the `data` type.